Re-introduce concept of projection cache 'completion'

Aaron1011 opened the pull request:
Instead of clearing out the cache entirely, we store
the intermediate evaluation result into the cache entry.
This accomplishes several things:

* We avoid the performance hit associated with re-evaluating
  the sub-obligations
* We avoid causing issues with incremental compilation, since
  the final evaluation result is always the same
* We avoid affecting other uses of the same `InferCtxt` which
  might care about 'side effects' from processing the sub-obligations
  (e,g. region constraints). Only code that is specifically aware
   of the new 'complete' code is affected

Aaron1011 commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit 58de3037235c48765dcdfb6530cb316ac025ec5b with merge fa1210416b5caa99212b19d594ef3da8ae3132f1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"58de3037235c48765dcdfb6530cb316ac025ec5b","merge_sha":"fa1210416b5caa99212b19d594ef3da8ae3132f1"} -->

rust-timer commented:
Finished benchmarking commit (fa1210416b5caa99212b19d594ef3da8ae3132f1): [comparison url](https://perf.rust-lang.org/compare.html?start=044674337a180c494b7e6fdce4b20dca93324b2a&end=fa1210416b5caa99212b19d594ef3da8ae3132f1).

**Summary**: This change led to very large relevant mixed results 🤷 in compiler performance.
- Very large improvement in instruction counts (up to -6.0% on `incr-unchanged` builds of `regression-31157`)
- Very large regression in instruction counts (up to 19.6% on `incr-unchanged` builds of `deeply-nested`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

nayato commented:
This fixes enormous regression we've seen introduced seemingly with https://github.com/rust-lang/rust/pull/85868.
Debug build before before #85868: 3 min
Same build after #85868: 12+ h - at which point we don't wait for it anymore and abort it.
Same build with this PR: 3.5 min

Aaron1011 commented:
@nayato: I'm glad to hear it! I came up with this as a fix for https://github.com/rust-lang/rust/issues/89195, which appears to have the same underlying cause as your issue.

Unfortunately, part of the issue with https://github.com/rust-lang/rust/issues/89195 seems to be be the 'eager' evaluation and discarding logic that I added in https://github.com/rust-lang/rust/pull/85868 (and that I've now removed from this PR). When there are a large number of projections with a large number of sub-obligations, this appears to blow up compilation time. Unfortunately, removing that logic appears to have a large *negative* impact on the `deeply-nested` benchmark.

bors commented:
:hourglass: Trying commit 6971dea4036a802fe9d7a838b8d3e655fee7d70f with merge 38dc5847c0733779aaaa9e3de4408165ea7a2b05...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6971dea4036a802fe9d7a838b8d3e655fee7d70f","merge_sha":"38dc5847c0733779aaaa9e3de4408165ea7a2b05"} -->

rust-timer commented:
Finished benchmarking commit (38dc5847c0733779aaaa9e3de4408165ea7a2b05): [comparison url](https://perf.rust-lang.org/compare.html?start=81117ff930fbf3792b4f9504e3c6bccc87b10823&end=38dc5847c0733779aaaa9e3de4408165ea7a2b05).

**Summary**: This change led to very large relevant mixed results 🤷 in compiler performance.
- Very large improvement in instruction counts (up to -6.0% on `incr-unchanged` builds of `regression-31157`)
- Very large regression in instruction counts (up to 19.8% on `incr-unchanged` builds of `deeply-nested`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

rust-timer commented:
Queued c80bb240f9760b0b0ed2e6045d60508129d31f26 with parent 5dab47dcd8267b8769421b46532414ec36d625e3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5dab47dcd8267b8769421b46532414ec36d625e3&end=c80bb240f9760b0b0ed2e6045d60508129d31f26).

JohnCSimon commented:
Returning to author because this seems to require more conversation. This also has a merge conflict.

Adjust the label with `@rustbot ready` when its ready for review

jackh726 commented:
After #91065 lands (soon), I'd like to start moving on getting this (and other related PRs) landed.

I think first, we should try to land the deduplication done in #84944/#90423. The only *additional* thing that #84944 does that #90423 does is remove the deduplication in `select/mod.rs`, which seems to provide a moderate perf gain (though, the base for these changes are different, so it might make no different). I'm also looking at the crater run done in #77325 and wondering why we never made tests for the variants found there (or did we?)

After that change lands (which should be a big perf win), we should take another look at this PR (see how perf looks). But, I'm inclined to agree with https://github.com/rust-lang/rust/pull/91183#issuecomment-981127777 that we likely just need to accept the perf hit.

However, I do think it would be super helpful for a general writeup of how projection and projection caching works. I don't think it needs to block this PR, but it would be really nice. There are issues like #80807 and #87755 that still lurk that it would be really nice to have a good explanation of the current system to help get a better birds-eye view of the problem, in case that helps us see a solution.

Aaron1011 commented:
Note that https://github.com/rust-lang/rust/pull/91186 is another PR that does some form of deduplication

Aaron1011 commented:
cc @rust-lang/infra: I don't know why `T-rustdoc` got added

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 314027bb8fe084f7a21268dc743505c0f4f5b2c5 with merge 35198c7a6d25cfbcbf61da23f8a6b22db99eacce...
<!-- homu: {"type":"TryBuildStarted","head_sha":"314027bb8fe084f7a21268dc743505c0f4f5b2c5","merge_sha":"35198c7a6d25cfbcbf61da23f8a6b22db99eacce"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4512755331?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/314027bb8fe084f7a21268dc743505c0f4f5b2c5/checks/4512755331/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_trait_selection/src/traits/project.rs at line 945:
             };
 
             let mut deduped: SsoHashSet<_> = Default::default();
-            result.obligations.drain_filter(|projected_obligation| {
-                !deduped.insert(projected_obligation.clone())
-            });
+                .obligations
+                .obligations
+                .drain_filter(|projected_obligation| !deduped.insert(projected_obligation.clone()));
             if use_cache {
             if use_cache {
                 infcx.inner.borrow_mut().projection_cache().insert_ty(cache_key, result.clone());
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs" "/checkout/compiler/rustc_trait_selection/src/traits/auto_trait.rs" "/checkout/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs" "/checkout/compiler/rustc_trait_selection/src/traits/project.rs" "/checkout/compiler/rustc_trait_selection/src/traits/relationships.rs" "/checkout/compiler/rustc_trait_selection/src/traits/chalk_fulfill.rs" "/checkout/compiler/rustc_trait_selection/src/traits/specialize/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4513602645?check_suite_focus=true)
Build commit: 35198c7a6d25cfbcbf61da23f8a6b22db99eacce (`35198c7a6d25cfbcbf61da23f8a6b22db99eacce`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4513602645?check_suite_focus=true"},"merge_sha":"35198c7a6d25cfbcbf61da23f8a6b22db99eacce"} -->

rust-timer commented:
Queued 35198c7a6d25cfbcbf61da23f8a6b22db99eacce with parent 1796de7bb123df3f3d32124ededf1344434f672e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1796de7bb123df3f3d32124ededf1344434f672e&end=35198c7a6d25cfbcbf61da23f8a6b22db99eacce).

rust-timer commented:
Finished benchmarking commit (35198c7a6d25cfbcbf61da23f8a6b22db99eacce): [comparison url](https://perf.rust-lang.org/compare.html?start=1796de7bb123df3f3d32124ededf1344434f672e&end=35198c7a6d25cfbcbf61da23f8a6b22db99eacce).

**Summary**: This change led to very large relevant mixed results 🤷 in compiler performance.
- Very large improvement in instruction counts (up to -5.9% on `incr-unchanged` builds of `regression-31157`)
- Very large regression in instruction counts (up to 8.6% on `incr-unchanged` builds of `deeply-nested`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

wesleywiser commented:
I retested on an internal crate that was experiencing a huge regression in build times with the 1.56 release:

- On nightly: 898 seconds to build
- This patch: 34 seconds to build

Really great work! 🎉 🎉 

I'm planning to review again tomorrow sometime and r+ assuming everything looks good.

jackh726 commented on compiler/rustc_infer/src/traits/project.rs:
```diff
@@ -149,10 +169,41 @@ impl<'tcx> ProjectionCache<'_, 'tcx> {
+                info!("ProjectionCacheEntry::complete({:?}) - completing {:?}", key, ty);
+                let mut ty = ty.clone();
+                if result == EvaluationResult::EvaluatedToOk {
+                    ty.obligations = vec![];
```
So, the docs above say "During evaluation, we will skip evaluating the cached sub-obligations in `ty` if this field is set." So, setting the obligations here to an empty vec is a bit redundant?

Seemingly, at the very least, the docs should be changed to reflect that the obligations are removed altogether. But maybe we can improve the type here to be encode the either-or state or "obligations or completion"

Aaron1011 replied:
I think an empty `Vec` is fine, and makes the type a little easier to work with. I'll update the docs to explain what's going on.

jackh726 commented on compiler/rustc_trait_selection/src/lib.rs:
```diff
@@ -12,11 +12,13 @@
 
 #![doc(html_root_url = "https://doc.rust-lang.org/nightly/nightly-rustc/")]
 #![feature(bool_to_option)]
+#![feature(backtrace)]
```
I think so too

jackh726 commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -550,8 +552,53 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                                            .inner
+                                            .borrow_mut()
+                                            .projection_cache()
+                                            .complete(key, res.unwrap());
```
I feel like you should be able to get around this `unwrap` with an if-let and binding...

jackh726 commented on compiler/rustc_trait_selection/src/traits/fulfill.rs:
```diff
@@ -709,6 +712,17 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
+                        .inner
+                        .borrow_mut()
+                        .projection_cache()
+                        .complete(key, EvaluationResult::EvaluatedToOk);
```
Currently we can `complete` a projection in two places. I'm curious which of the two gives the perf benefits. It could be that one is essentially useless (in which case that might suggest that code path is not followed much - at least in terms where caching could be helpful).

Either way, I think I'd at least like some better comments about *why* we can cache at each place.

jackh726 commented on compiler/rustc_trait_selection/src/traits/project.rs:
```diff
@@ -946,27 +945,8 @@ fn opt_normalize_projection_type<'a, 'b, 'tcx>(
-                // the result to `EvaluatedToOkModuloRegions`, while an
-                // `EvaluatedToOk` obligation will never change the result.
-                // See #85360 for more details
-                projected_obligation.is_global(canonical.tcx())
-                    && canonical
-                        .evaluate_root_obligation(projected_obligation)
-                        .map_or(false, |res| res.must_apply_considering_regions())
```
I think I'd like to see this change pulled out to a separate PR - considering it has it's own bug fixes. Or at least have a separate PR filed to see the perf effects of removing this on its own.

jackh726 reviewed (changes requested):
I have some things I'd like to see here for perf experimenting - but that doesn't necessarily need to happen.

There are some comments that I'd like changed though, after that r=me if @Aaron1011 doesn't want to do the perf experiments I commented on.

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit a68194bc06d0c5121e18266330b8835b7d9dee73 with merge 52f485bc100595b3ce1a438e26db26e82ac3f65f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a68194bc06d0c5121e18266330b8835b7d9dee73","merge_sha":"52f485bc100595b3ce1a438e26db26e82ac3f65f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4563752480?check_suite_focus=true)
Build commit: 52f485bc100595b3ce1a438e26db26e82ac3f65f (`52f485bc100595b3ce1a438e26db26e82ac3f65f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4563752480?check_suite_focus=true"},"merge_sha":"52f485bc100595b3ce1a438e26db26e82ac3f65f"} -->

rust-timer commented:
Queued 52f485bc100595b3ce1a438e26db26e82ac3f65f with parent 2595d038273a4c7746320efd06200764f99bd760, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2595d038273a4c7746320efd06200764f99bd760&end=52f485bc100595b3ce1a438e26db26e82ac3f65f).

Mark-Simulacrum commented:
Note that due to merging https://github.com/rust-lang/rustc-perf/pull/1123, these runs may look quite bad -- once we get another master commit landed + benchmarked, re-running perf should make any noise there go away. It shouldn't affect non-incremental results, though, of course.

(Leaving this comment on all pending perf run PRs).

rust-timer commented:
Finished benchmarking commit (52f485bc100595b3ce1a438e26db26e82ac3f65f): [comparison url](https://perf.rust-lang.org/compare.html?start=2595d038273a4c7746320efd06200764f99bd760&end=52f485bc100595b3ce1a438e26db26e82ac3f65f).

**Summary**: This change led to very large relevant regressions 😿 in compiler performance.
- Very large regression in instruction counts (up to 133.3% on `incr-unchanged` builds of `ctfe-stress-4`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 2517bd2afe2eae4fadbaf8baf230110ae7957026 with merge 02aeffa72b11c0501c63a17d3e3f27c833ef2cbf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2517bd2afe2eae4fadbaf8baf230110ae7957026","merge_sha":"02aeffa72b11c0501c63a17d3e3f27c833ef2cbf"} -->

jackh726 commented:
r=me with okay perf

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4571111691?check_suite_focus=true)
Build commit: 02aeffa72b11c0501c63a17d3e3f27c833ef2cbf (`02aeffa72b11c0501c63a17d3e3f27c833ef2cbf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4571111691?check_suite_focus=true"},"merge_sha":"02aeffa72b11c0501c63a17d3e3f27c833ef2cbf"} -->

rust-timer commented:
Queued 02aeffa72b11c0501c63a17d3e3f27c833ef2cbf with parent d3848cb659105d8384c00ba384ec2f6f39c47238, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d3848cb659105d8384c00ba384ec2f6f39c47238&end=02aeffa72b11c0501c63a17d3e3f27c833ef2cbf).

rust-timer commented:
Finished benchmarking commit (02aeffa72b11c0501c63a17d3e3f27c833ef2cbf): [comparison url](https://perf.rust-lang.org/compare.html?start=d3848cb659105d8384c00ba384ec2f6f39c47238&end=02aeffa72b11c0501c63a17d3e3f27c833ef2cbf).

**Summary**: This change led to small relevant mixed results 🤷 in compiler performance.
- Small improvement in instruction counts (up to -1.2% on `full` builds of `regression-31157`)
- Small regression in instruction counts (up to 0.5% on `incr-unchanged` builds of `webrender-wrench`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

Aaron1011 commented:
@bors r=jackh726

bors commented:
:pushpin: Commit 40ef1d322304c8a21c675fd32886fb27ebe07039 has been approved by `jackh726`

<!-- @bors r=jackh726 40ef1d322304c8a21c675fd32886fb27ebe07039 -->
<!-- homu: {"type":"Approved","sha":"40ef1d322304c8a21c675fd32886fb27ebe07039","approver":"jackh726"} -->

bors commented:
:hourglass: Testing commit 40ef1d322304c8a21c675fd32886fb27ebe07039 with merge d6cffe41b59feaab5fb92bb320e60586202c9950...
<!-- homu: {"type":"BuildStarted","head_sha":"40ef1d322304c8a21c675fd32886fb27ebe07039","merge_sha":"d6cffe41b59feaab5fb92bb320e60586202c9950"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4573130143?check_suite_focus=true)
Approved by: jackh726
Pushing d6cffe41b59feaab5fb92bb320e60586202c9950 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4573130143?check_suite_focus=true"},"merge_sha":"d6cffe41b59feaab5fb92bb320e60586202c9950"} -->

rust-timer commented:
Finished benchmarking commit (d6cffe41b59feaab5fb92bb320e60586202c9950): [comparison url](https://perf.rust-lang.org/compare.html?start=df2f45c1d30615c930b7aa76bb9efbeea3909703&end=d6cffe41b59feaab5fb92bb320e60586202c9950).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

matze commented:
Any chance this can make it into 1.58? It fixes severe build times we have since 1.56 in a tonic/tower-based application of ours and essentially holding us back from upgrading to edition 2021 and a more recent rustls.

jackh726 commented:
I'll nominate this for stable and beta.

Personally, I think is *at least* as risky as #90423, and that got stable declined. But I'll beta backport might be okay here. I'll nominate both for discussion.

pnkfelix commented:
Discussed in T-compiler meeting. Deferring decision on whether  to backport (any of beta/stable) until next week, at least.

apiraino commented:
Discussed again today and the decision is to not baclport this on both beta or stable channels [Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-12-30.20.2354818/near/266435406)

@rustbot label -beta-nominated -stable-nominated
